File: local/mongoConnector.py
import pymongo
import os

class MongoDB:
    def __init__(self):
        
        uri = os.environ["MONGODB_URI"]
        self.client = pymongo.MongoClient(uri)
        print("MongoConnector: Connected to MongoDB")
        self.db = self.client["PiGen"]
        self.user_collection = self.db["users"]
        self.setting_collection = self.db["settings"]

    def check_admin(self, username):
        user = self.user_collection.find_one({"username": username})
        return user["admin"] if user else False

    def check_user(self, user):
        if not user:
            return False
        db_user = self.user_collection.find_one({"username": user["username"]})
        if not db_user:
            return False
        return self.validate_password(user["username"], user["password"], False)

    def user_exists(self, username):
        user = self.user_collection.find_one({"username": username})
        return user is not None

    def get_user_password(self, username):
        user = self.user_collection.find_one({"username": username})
        return user["password"] if user else None

    def get_user_admin(self, username):
        user = self.user_collection.find_one({"username": username})
        return user["admin"] if user else False

    def delete_user(self, username):
        self.user_collection.delete_one({"username": username})

    def create_user(self, username, password, admin):
        new_user = {"username": username, "password": password, "admin": admin}
        self.user_collection.insert_one(new_user)

    def list_users(self):
        users = self.user_collection.find({}, {"username": 1, "_id": 0})
        return [user["username"] for user in users]

    def validate_password(self, username, key, log=True):
        password = self.get_user_password(username)
        if password:
            if password == key:
                if log:
                    print(f"MongoConnector: User '{username}' is authenticated.")
                return True
            else:
                if log:
                    print("MongoConnector: Password is incorrect")
                return f"Invalid password (name {username} Pass: {key})"
        else:
            if log:
                print(f"MongoConnector: User does not exist (name {username} Pass: {key})")
            return "This User does not exist."

    def set_setting(self, setting, value):
        existing_setting = self.setting_collection.find_one({"settingName": setting})
        if existing_setting:
            existing_setting["settingValue"] = value
            self.setting_collection.update_one({"_id": existing_setting["_id"]}, {"$set": existing_setting})
        else:
            new_setting = {"settingName": setting, "settingValue": value}
            self.setting_collection.insert_one(new_setting)

    def get_setting(self, setting):
        found_setting = self.setting_collection.find_one({"settingName": setting})
        return found_setting["settingValue"] if found_setting else None

    def set_user_data(self, username, key, value):
        user = self.user_collection.find_one({"username": username})
        if not user:
            new_user = {"username": username, "userData": [{"key": key, "value": value}]}
            self.user_collection.insert_one(new_user)
        else:
            user_data = next((data for data in user["userData"] if data["key"] == key), None)
            if not user_data:
                user["userData"].append({"key": key, "value": value})
                self.user_collection.update_one({"_id": user["_id"]}, {"$set": user})
            else:
                user_data["value"] = value
                self.user_collection.update_one({"_id": user["_id"], "userData.key": key}, {"$set": {"userData.$": user_data}})

    def get_user_data(self, username, key):
        user = self.user_collection.find_one({"username": username})
        if user:
            user_data = next((data for data in user["userData"] if data["key"] == key), None)
            return user_data["value"] if user_data else None
        return None

File: server/MongoConnector.js
module.exports.MongoDB = class MongoDB {
    constructor() {
        this.mongoose = require('mongoose');
        const Schema = this.mongoose.Schema;

        const userDataSchema = new Schema({
            key: { type: String, required: true },
            value: { type: Object, required: true }
        });

        const userSchema = new Schema({
            username: { type: String, required: true, unique: true },
            password: { type: String, required: true },
            admin: { type: Boolean, required: true },
            userData: [userDataSchema]
        });

        this.user = this.mongoose.model('User', userSchema);

        const settingsSchema = new Schema({
            settingName: { type: String, required: true },
            settingValue: { type: String, required: true }
        });

        this.setting = this.mongoose.model('Setting', settingsSchema);

        const uri = process.env.MONGODB_URI;


        this.mongoose.connect(uri)
            .then(() => console.log('MongoDB connection successful'))
            .catch(err => console.error('MongoDB connection error:', err));
    }

    async checkAdmin(username) {
        const user = await this.user.findOne({ username: username });
        return user ? user.admin : false;
    }

    async checkUser(user) {
        if (!user) {
            return false;
        }
        const dbUser = await this.user.findOne({ username: user.username });
        if (!dbUser) {
            return false;
        }
        return this.validate_password(user.username, user.password, false);
    }

    async userExists(username) {
        const user = await this.user.findOne({ username: username });
        return user != null;
    }

    async getUserPassword(username) {
        const user = await this.user.findOne({ username: username });
        return user ? user.password : null;
    }

    async getUserAdmin(username) {
        const user = await this.user.findOne({ username: username });
        return user ? user.admin : false;
    }

    async deleteUser(username) {
        await this.user.deleteOne({ username: username });
    }

    async createUser(username, password, admin) {
        const hashedPassword = this.hashPassword(password);
        const existingUser = await this.user.findOne({ username: username });
        if (existingUser) {
            existingUser.password = hashedPassword;
            existingUser.admin = admin;
            await existingUser.save();
        } else {
            const newUser = new this.user({ username, password: hashedPassword, admin });
            await newUser.save();
        }
    }

    async listUsers() {
        return await this.user.find({}, 'username -_id').then(users => users.map(user => user.username));
    }

    async validate_password(username, key, log = true) {
        const hashedPassword = this.hashPassword(key);
        const password = await this.getUserPassword(username);
        if (password) {
            if (password == hashedPassword) {
                if (log)
                    console.log("MongoConnector: User '" + username + "' is authenticated.");
                return true;
            }
            else {
                if (log)
                    console.log("MongoConnector: Password is incorrect");
                return false;
            }
        }
        else {
            if (log)
                console.log("MongoConnector: User does not exist (name " + username + " Pass: " + hashedPassword + ")");
            return false;
        }
    }

    async setSetting(setting, value) {
        const existingSetting = await this.setting.findOne({ settingName: setting });
        if (existingSetting) {
            existingSetting.settingValue = value;
            await existingSetting.save();
        } else {
            const newSetting = new this.setting({ settingName: setting, settingValue: value });
            await newSetting.save();
        }
    }

    async getSetting(setting) {
        const foundSetting = await this.setting.findOne({ settingName: setting });
        return foundSetting ? foundSetting.settingValue : null;
    }



    async setUserData(username, key, value) {
        const user = await this.user.findOne({ username: username });
        if (!user) {
            const newUser = new this.user({ username: username });
            newUser.userData.push({ key: key, value: value });
            await newUser.save();
        } else {
            if (!user.userData) {
                user.userData = [];
            }
            const userData = user.userData.find(data => data.key === key);
            if (!userData) {
                user.userData.push({ key: key, value: value });
                await user.save();
            } else {
                userData.value = value;
                await user.save();
            }
        }
    }
    
    async getUserData(username, key) {
        const user = await this.user.findOne({ username: username });
        if (user.userData) {
            const userData = user.userData.find(data => data.key === key);
            return userData ? userData.value : null;
        }
        return null;
    }

    async checkConnection() {
        return this.mongoose.connection.readyState;
    }

    hashPassword(password) {
        const crypto = require('crypto');
        const hash = crypto.createHash('sha256');
        hash.update(password);
        return hash.digest('hex');
    }
    
}